common/src/main/java/com/linecorp/centraldogma/internal/api/v1/ProjectDto.java

```diff
@@ -24,36 +24,58 @@
 
 import javax.annotation.Nullable;
 
+import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonInclude;
 import com.fasterxml.jackson.annotation.JsonInclude.Include;
 import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.MoreObjects;
 import com.google.common.base.MoreObjects.ToStringHelper;
 
 import com.linecorp.centraldogma.common.Author;
+import com.linecorp.centraldogma.common.ProjectRole;
 
 @JsonInclude(Include.NON_NULL)
 public class ProjectDto {
 
     private final String name;
 
+    @Nullable
     private Author creator;
 
+    @Nullable
     private String url;
 
-    private String reposUrl;
+    @Nullable
+    private ProjectRole userRole;
 
+    @Nullable
     private String createdAt;
 
     public ProjectDto(String name) {
         this.name = requireNonNull(name, "name");
     }
 
-    public ProjectDto(String name, Author creator, long creationTimeMillis) {
+    @VisibleForTesting
+    @JsonCreator
+    public ProjectDto(@JsonProperty("name") String name,
+                      @JsonProperty("creator") @Nullable Author creator,
+                      @JsonProperty("url") @Nullable String url,
+                      @JsonProperty("userRole") @Nullable ProjectRole userRole,
+                      @JsonProperty("createdAt") @Nullable String createdAt) {
```

Comment:
Question) Even in tests, is there a need for the nullable contract since server-side always seems to fill in these fields?

Reply:
This constructor was added for deserializing the API responses in tests.
https://github.com/line/centraldogma/blob/11e03be535ec2f0c3bd942f161169f8b7af24eb4/server/src/test/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1ListProjectTest.java#L339-L340

It wasn't used for `?status=removed` responses but made them nullable for future usage.

Reply:
> This constructor was added for deserializing the API responses in tests.

I understand that. I meant even in the test you linked aren't the fields always being filled in (non-null) anyways?

Reply:
> you linked aren't the fields always being filled in (non-null) anyways?

Right. They are non-null for normal projects. The values are nullable for removed projects.
https://github.com/line/centraldogma/blob/c0673c169c9a19034719c5a0e2594bc85415a577/server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java#L84-L86